hoomd/hpmc/ExternalFieldHarmonic.h

```diff
@@ -0,0 +1,482 @@
+
+#include "ExternalField.h"
+
+#ifndef __HIPCC__
+#include <pybind11/pybind11.h>
+#endif
+
+namespace hoomd
+    {
+namespace hpmc
+    {
+
+template<class Shape> class ExternalFieldHarmonic : public ExternalFieldMono<Shape>
+    {
+    using ExternalFieldMono<Shape>::m_pdata;
+    using ExternalFieldMono<Shape>::m_exec_conf;
+    using ExternalFieldMono<Shape>::m_sysdef;
+
+    public:
+    //! Constructor
+    ExternalFieldHarmonic(std::shared_ptr<SystemDefinition> sysdef,
+                          pybind11::array_t<double> r0,
+                          Scalar k,
+                          pybind11::array_t<double> q0,
+                          Scalar q,
+                          pybind11::array_t<double> symRotations)
+        : ExternalFieldMono<Shape>(sysdef), m_k_translational(k), m_k_rotational(q)
+        {
+        setReferencePositions(r0);
+        setReferenceOrientations(q0);
+        setSymmetricallyEquivalentOrientations(symRotations); // TODO: check for identity?
+
+        // connect updateMemberTags() method to maximum particle number change signal
+        m_pdata->getGlobalParticleNumberChangeSignal()
+            .template connect<ExternalFieldHarmonic,
+                              &ExternalFieldHarmonic::slotGlobalParticleNumChange>(this);
+        } // end constructor
+
+    //! Destructor
+    ~ExternalFieldHarmonic()
+        {
+        if (m_pdata)
+            {
+            m_pdata->getGlobalParticleNumberChangeSignal()
+                .template disconnect<ExternalFieldHarmonic,
+                                     &ExternalFieldHarmonic::slotGlobalParticleNumChange>(this);
+            }
+        } // end destructor
+
+    //! Set reference positions from a (N_particles, 3) numpy array
+    void setReferencePositions(const pybind11::array_t<double> ref_pos)
+        {
+        m_reference_positions.resize(m_pdata->getNGlobal());
+        if (m_exec_conf->getRank() == 0)
+            {
+            if (ref_pos.ndim() != 2)
+                {
+                throw std::runtime_error("The array must be of shape (N_particles, 3).");
+                }
+
+            const size_t N_particles = ref_pos.shape(0);
+            const size_t dim = ref_pos.shape(1);
+            if (N_particles != m_pdata->getNGlobal() || dim != 3)
+                {
+                throw std::runtime_error("The array must be of shape (N_particles, 3).");
+                }
+            const double* rawdata = static_cast<const double*>(ref_pos.data());
+            for (size_t i = 0; i < std::min<unsigned long>(N_particles, m_pdata->getNGlobal()); i++)
+                {
+                const size_t array_index = i * 3;
+                this->m_reference_positions[i] = vec3<Scalar>(rawdata[array_index],
+                                                              rawdata[array_index + 1],
+                                                              rawdata[array_index + 2]);
+                }
+            }
```

Comment:
```suggestion
            for (size_t i = 0; i < N_particles; i+=3)
                {
                this->m_reference_positions[i] = vec3<Scalar>(rawdata[i]],
                                                              rawdata[i + 1],
                                                              rawdata[i + 2]);
                }
            }
```

Reply:
This can be applied to the other loops for this purpose.

Reply:
> This can be applied to the other loops for this purpose.

...with appropriate modifications of course (: